File: numethods/interpolation/interpolation.m
function interpolation()
    % Устанавливаем формат отображения чисел
    format long g;
    
    % Задаём количество узлов и концы интервала
    N = 5;
    a = -3/2;
    b = 3/2;
    h = (b - a) / (N - 1);
    
    % Равноудалённые узлы
    nodes1 = a:h:b;
    
    % Узлы по Чебышёву
    nodes2 = zeros(1, N);
    for i = 0:(N-1)
        nodes2(i+1) = 0.5 * ((b - a)*cos((2*i + 1)*pi/(2*N)) + (b + a));
    end
    
    % Построение интерполяционных полиномов Лагранжа с использованием встроенных операций
    L1 = LagrangePolynomial(nodes1);
    L2 = LagrangePolynomial(nodes2);
    
    fprintf('Количество узлов: %d\n\n', N);
    
    % Вывод полинома для равноудалённых узлов
    fprintf('Полином по равноудалённым узлам:\nL1(x) = ');
    deg = length(L1)-1;
    for i = 1:length(L1)-1
        fprintf('%.9f*x^%d + ', L1(i), deg);
        deg = deg - 1;
    end
    fprintf('%.9f\n\n', L1(end));
    
    % Вывод полинома для узлов Чебышёва
    fprintf('Полином по узлам, определяемым полиномами Чебышёва:\nL2(x) = ');
    deg = length(L2)-1;
    for i = 1:length(L2)-1
        fprintf('%.9f*x^%d + ', L2(i), deg);
        deg = deg - 1;
    end
    fprintf('%.9f\n\n', L2(end));
    
    % Вывод значений полиномов и функции f(x) в случайных точках
    fprintf('x\t\tL1(x)\t\tL2(x)\t\tf(x)\n');
    for i = 1:10
        % Генерируем случайное x в диапазоне [-1.5, 1.5]
        x = (randi([0,3000])-1500)/1000;
        valL1 = polyval(L1, x);
        valL2 = polyval(L2, x);
        valF = f(x);
        fprintf('%.4f\t%.9f\t%.9f\t%.9f\n', x, valL1, valL2, valF);
    end
end

% Функция f: f(x) = x - sin(x) - 0.25
function y = f(x)
    y = x - sin(x) - 0.25;
end

% Функция для сложения двух полиномов, представленных в виде векторов коэффициентов.
function sumPoly = polyAdd(p1, p2)
    % Приводим полиномы к одинаковой длине, дополняя нулями слева
    d = max(length(p1), length(p2));
    p1 = [zeros(1, d - length(p1)) p1];
    p2 = [zeros(1, d - length(p2)) p2];
    sumPoly = p1 + p2;
end

% Построение интерполяционного полинома Лагранжа по заданным узлам.
% Возвращает вектор коэффициентов [a_n, ..., a_0] для полинома a_n*x^n + ... + a_0.
function P = LagrangePolynomial(nodes)
    N = length(nodes);
    L_res = 0;  % Инициализируем результирующий полином нулём
    % Для каждого узла строим соответствующий базисный полином
    for i = 1:N
        basis = 1;  % Начинаем с полинома 1
        for j = 1:N
            if i ~= j
                % Используем встроенную функцию conv для умножения полиномов:
                % умножаем basis на (x - nodes(j))
                basis = conv(basis, [1, -nodes(j)]);
                % Нормируем множитель, делим коэффициенты на (nodes(i) - nodes(j))
                basis = basis / (nodes(i) - nodes(j));
            end
        end
        % Умножаем базисный многочлен на значение f в узле nodes(i)
        basis = basis * f(nodes(i));
        % Складываем полученные базисные многочлены с помощью polyAdd
        L_res = polyAdd(L_res, basis);
    end
    P = L_res;
end

File: numethods/interpolation/interpolation.cpp
#include <bits/stdc++.h>

using namespace std;
using ld = long double;
const ld PI = 3.14159265f;

ld f(ld x) {
    return x - sin(x) - 0.25f;
}

vector<ld> polynomialMultiplication(const vector<ld>& p1, const vector<ld>& p2) {
    int deg1 = p1.size() - 1;
    int deg2 = p2.size() - 1;
    vector<ld> p(deg1 + deg2 + 1, 0.0f);
    for (int i = 0; i <= deg1; ++i) {
        for (int j = 0; j <= deg2; ++j) {
            p[i + j] += p1[i] * p2[j];
        }
    }
    return p;
}

ld polynomialCalculation(const vector<ld>& p, ld x) {
    ld result = 0.0f;
    int degree = p.size();
    for (int i = 0; i < degree; ++i) {
        result += p[i] * pow(x, degree - 1 - i);
    }
    return result;
}

vector<ld> polynomialAddition(const vector<ld>& p1, const vector<ld>& p2) {
    // Приводим полиномы к одинаковой длине, дополняя нулями в начале
    size_t size1 = p1.size();
    size_t size2 = p2.size();
    size_t maxSize = max(size1, size2);

    vector<ld> poly1(maxSize - size1, 0.0f);
    poly1.insert(poly1.end(), p1.begin(), p1.end());
    vector<ld> poly2(maxSize - size2, 0.0f);
    poly2.insert(poly2.end(), p2.begin(), p2.end());

    vector<ld> sum(maxSize, 0.0f);
    for (size_t i = 0; i < maxSize; ++i) {
        sum[i] = poly1[i] + poly2[i];
    }
    return sum;
}

vector<ld> LagrangePolynomial(const vector<ld>& nodes) {
    int N = nodes.size();
    vector<ld> L(N, 0.0f);
    vector<vector<ld>> basis(N, vector<ld>(1, 1.0f));

    for (int i = 0; i < N; ++i) {
        for (int j = 0; j < N; ++j) {
            if (i != j) {
                // Домножаем на (x - nodes[j])
                basis[i] = polynomialMultiplication(basis[i], {1.0f, -nodes[j]});
                // Нормируем множитель
                ld factor = 1.0f / (nodes[i] - nodes[j]);
                for (auto &coef : basis[i]) {
                    coef *= factor;
                }
            }
        }
        // Умножаем базисный многочлен на f(nodes[i])
        basis[i] = polynomialMultiplication(basis[i], {f(nodes[i])});
    }

    // Складываем полученные базисные многочлены
    for (int i = 0; i < N; ++i) {
        L = polynomialAddition(L, basis[i]);
    }
    return L;
}

int main() {
    cout << setprecision(9) << fixed;
    const int N = 5; // количество узлов
    const ld a = -PI, b = PI; // концы интервала
    const ld h = (b - a) / (N - 1); // шаг

    vector<ld> nodes1(N);
    for (int i = 0; i < N; ++i) {
        nodes1[i] = a + h * i;
    }

    vector<ld> nodes2(N);
    for (int i = 0; i < N; ++i) {
        nodes2[i] = 0.5f * ((b - a) * cos((2 * i + 1) * PI / (2 * N)) + (b + a));
    }

    // Построение интерполяционных полиномов
    vector<ld> L1 = LagrangePolynomial(nodes1);
    vector<ld> L2 = LagrangePolynomial(nodes2);

    cout << "Количество узлов: " << N << "\n\n";

    cout << "Полином по равноудаленным узлам:\nL1(x) = ";
    for (int i = 0; i < N - 1; ++i) {
        cout << L1[i] << "x^" << N - i - 1 << " + ";
    }
    cout << L1[N - 1] << "\n\n";

    cout << "Полином по узлам, определяемым полиномами Чебышева:\nL2(x) = ";
    for (int i = 0; i < N - 1; ++i) {
        cout << L2[i] << "x^" << N - i - 1 << " + ";
    }
    cout << L2[N - 1] << "\n\n";

    cout << "x\t\tL1(x)\t\tL2(x)\t\tf(x)\n";
    for (int i = 0; i < 10; ++i) {
        ld x = (rand() % 3001 - 1500) / 1000.0f;
        cout << x << "\t" 
             << f(x) << "\t" 
             << polynomialCalculation(L2, x) << "\t" 
             << polynomialCalculation(L1, x) << "\n";
    }
    return 0;
}